Add Linux support

agisboye opened the pull request:
This adds a little bit of conditional compilation so that Wrap will build on Linux.

agisboye committed:
Add Linux support

JohnSundell commented on Sources/Wrap.swift:
```diff
@@ -261,7 +261,11 @@ public extension WrapCustomizable {
     func convertPropertyNameToSnakeCase(propertyName: String) -> String {
-        let regex = try! NSRegularExpression(pattern: "(?<=[a-z])([A-Z])|([A-Z])(?=[a-z])", options: [])
+        #if os(OSX) || os(iOS) || os(watchOS) || os(tvOS)
+            let regex = try! NSRegularExpression(pattern: "(?<=[a-z])([A-Z])|([A-Z])(?=[a-z])", options: [])
```
How about adding a typealias instead, to avoid having conditional compilation inline? This could be added at the top of the file:

```swift
#if !os(Linux)
typealias RegularExpression = NSRegularExpression
#endif
```

agisboye replied:
It'll work for now. 😄 I'd imagine that the value type `RegularExpression` will become available on Apple platforms, at which point the conditional should be removed. My other comment is applicable here as well.

JohnSundell commented on Sources/Wrap.swift:
```diff
@@ -340,11 +344,13 @@ extension NSArray: WrapCustomizable {
-extension NSDictionary: WrapCustomizable {
-    public func wrap(context: Any?, dateFormatter: DateFormatter?) -> Any? {
-        return try? Wrapper(context: context, dateFormatter: dateFormatter).wrap(dictionary: self as [NSObject : AnyObject])
+#if os(OSX) || os(iOS) || os(watchOS) || os(tvOS)
```
Let's do `#if !os(Linux)` instead, to make this a bit simpler 😄

agisboye replied:
I'll admit its simpler, but its also more fragile. Eventually (fingers crossed) Swift will be available on other platforms than Apple & Linux (think Windows, FreeBSD, Android etc.). I try to make my code future proof, hoping that Swift will be available on other platforms sooner rather than later.

JohnSundell replied:
The same argument can be made for if Apple introduces a new platform, then this code also needs to be updated. I don't think there's a way to make it 100% futureproof, so I think we should go for the simpler option for now 😄

agisboye replied:
Good point. 😉  Let me know if additional changes are needed. Otherwise I'd be great if you could add a new tag/release, so that SPM can fetch the latest version.

JohnSundell commented on Sources/Wrap.swift:
```diff
@@ -340,11 +344,13 @@ extension NSArray: WrapCustomizable {
-    public func wrap(context: Any?, dateFormatter: DateFormatter?) -> Any? {
-        return try? Wrapper(context: context, dateFormatter: dateFormatter).wrap(dictionary: self as [NSObject : AnyObject])
+#if os(OSX) || os(iOS) || os(watchOS) || os(tvOS)
+    extension NSDictionary: WrapCustomizable {
```
Don't change the indentiation here.

JohnSundell commented on Sources/Wrap.swift:
```diff
@@ -361,11 +367,13 @@ extension Date: WrappableDate {
-extension NSDate: WrappableDate {
-    public func wrap(dateFormatter: DateFormatter) -> String {
-        return dateFormatter.string(from: self as Date)
+#if os(OSX) || os(iOS) || os(watchOS) || os(tvOS)
```
Same feedback as above, use a simpler conditional, and unstage the indentation changes.

JohnSundell reviewed (changes requested):
Thanks for this @agisboye! So awesome to get Wrap up and running on Linux 😄 Have some inline suggestions, let me know what you think 🙂

JohnSundell committed:
Modernize repo & make it fully SPM compatible

This change makes the Wrap project a bit more modern, by re-generating the
project files using SwiftPlate, and fixing a compiler error when using SPM
to test the project. The project is now up-to-date and fully SPM compatible.

JohnSundell commented:
@agisboye Would love to get this merged! Do you mind making the modifications I suggested? 🙂

agisboye committed:
Simplify conditional compilation check

JohnSundell commented:
@agisboye here seems to be quite a lot of unrelated changes in here. Can you base this PR on the latest master? 🙂

agisboye committed:
Revert "Modernize repo & make it fully SPM compatible"

This reverts commit 434fc4f92f6887563be8b256cfcbeeeac07a83fd.

agisboye committed:
Merge remote-tracking branch 'JohnSundell/master'

JohnSundell reviewed (approved):
Awesome! Thanks a lot @agisboye! 🚀
